From 3288daa2c885da49190a013a5dcd8688d08c1fcf Mon Sep 17 00:00:00 2001
From: Zhan Lou <zhan.lou@cixtech.com>
Date: Fri, 28 Apr 2023 14:09:58 +0800
Subject: [PATCH] DPTSW-4457: v4l2: enable AV1 decoder element

Enable V4L2 AV1 basic decoder element, and pass test with NULL HW
driver.

Change-Id: Iebca673646b4426e14d27c7ab207c3d71de268ce
Signed-off-by: Zhan Lou <zhan.lou@cixtech.com>

diff --git a/sys/v4l2/gstv4l2av1codec.c b/sys/v4l2/gstv4l2av1codec.c
new file mode 100644
index 0000000000..107546b28d
--- /dev/null
+++ b/sys/v4l2/gstv4l2av1codec.c
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2023 Cix Technology (Shanghai) Co., Ltd.
+ *    Author: Zhan Lou <zhan.lou@cixtech.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "gstv4l2av1codec.h"
+#include "mvx-v4l2-controls.h"
+
+#include <gst/gst.h>
+#include "ext/v4l2-controls.h"
+
+
+static gint
+v4l2_profile_from_string (const gchar * profile)
+{
+  gint v4l2_profile = -1;
+
+  if (g_str_equal (profile, "0"))
+    v4l2_profile = 0;
+  else if (g_str_equal (profile, "1"))
+    v4l2_profile = 1;
+  else if (g_str_equal (profile, "2"))
+    v4l2_profile = 2;
+  else
+    GST_WARNING ("Unsupported profile string '%s'", profile);
+
+  return v4l2_profile;
+}
+
+static const gchar *
+v4l2_profile_to_string (gint v4l2_profile)
+{
+  switch (v4l2_profile) {
+    case 0:
+      return "0";
+    case 1:
+      return "1";
+    case 2:
+      return "2";
+    default:
+      GST_WARNING ("Unsupported V4L2 profile %i", v4l2_profile);
+      break;
+  }
+
+  return NULL;
+}
+
+const GstV4l2Codec *
+gst_v4l2_av1_get_codec (void)
+{
+  static GstV4l2Codec *codec = NULL;
+  if (g_once_init_enter (&codec)) {
+    static GstV4l2Codec c;
+    c.profile_cid = V4L2_CID_MVE_VIDEO_AV1_PROFILE;
+    c.profile_to_string = v4l2_profile_to_string;
+    c.profile_from_string = v4l2_profile_from_string;
+    g_once_init_leave (&codec, &c);
+  }
+  return codec;
+}
diff --git a/sys/v4l2/gstv4l2av1codec.h b/sys/v4l2/gstv4l2av1codec.h
new file mode 100644
index 0000000000..879b623af2
--- /dev/null
+++ b/sys/v4l2/gstv4l2av1codec.h
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2023 Cix Technology (Shanghai) Co., Ltd.
+ *    Author: Zhan Lou <zhan.lou@cixtech.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef __GST_V4L2_AV1_CODEC_H__
+#define __GST_V4L2_AV1_CODEC_H__
+
+#include "gstv4l2codec.h"
+
+G_BEGIN_DECLS
+
+const GstV4l2Codec * gst_v4l2_av1_get_codec (void);
+
+G_END_DECLS
+
+#endif
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 80189d28d0..1bc79409f9 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -235,6 +235,7 @@ static GstV4L2FormatDesc gst_v4l2_formats[] = {
   {MAP_ENC_FMT (VC1_ANNEX_L, ENCODED),  GST_V4L2_CODEC},
   {MAP_ENC_FMT (VP8, ENCODED),          GST_V4L2_CODEC | GST_V4L2_NO_PARSE},
   {MAP_ENC_FMT (VP9, ENCODED),          GST_V4L2_CODEC | GST_V4L2_NO_PARSE},
+  {MAP_ENC_FMT (AV1, ENCODED),          GST_V4L2_CODEC | GST_V4L2_NO_PARSE},
 
   /*  Vendor-specific formats   */
   {MAP_ENC_FMT (WNVA, ENCODED),     GST_V4L2_CODEC | GST_V4L2_RESOLUTION_AND_RATE},
@@ -1592,6 +1593,9 @@ gst_v4l2_object_v4l2fourcc_to_bare_struct (guint32 fourcc,
     case V4L2_PIX_FMT_VP9:
       structure = gst_structure_new_empty ("video/x-vp9");
       break;
+    case V4L2_PIX_FMT_AV1:
+      structure = gst_structure_new_empty ("video/x-av1");
+      break;
     case V4L2_PIX_FMT_DV:
       structure =
           gst_structure_new ("video/x-dv", "systemstream", G_TYPE_BOOLEAN, TRUE,
@@ -2041,6 +2045,8 @@ gst_v4l2_object_get_caps_info (GstV4l2Object * v4l2object, GstCaps * caps,
     fourcc = V4L2_PIX_FMT_VP8;
   } else if (g_str_equal (mimetype, "video/x-vp9")) {
     fourcc = V4L2_PIX_FMT_VP9;
+  } else if (g_str_equal (mimetype, "video/x-av1")) {
+    fourcc = V4L2_PIX_FMT_AV1;
   } else if (g_str_equal (mimetype, "video/x-bayer")) {
     const gchar *format = gst_structure_get_string (structure, "format");
     if (format) {
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index b7cd9ace7e..cdb0f7bbbc 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -38,6 +38,7 @@
 #include "gstv4l2mpeg4codec.h"
 #include "gstv4l2vp8codec.h"
 #include "gstv4l2vp9codec.h"
+#include "gstv4l2av1codec.h"
 
 #include <string.h>
 #include <glib/gi18n-lib.h>
@@ -1451,6 +1452,9 @@ G_STMT_START { \
   } else if (gst_structure_has_name (s, "video/x-vp9")) {
     SET_META ("VP9");
     cdata->codec = gst_v4l2_vp9_get_codec ();
+  } else if (gst_structure_has_name (s, "video/x-av1")) {
+    SET_META ("AV1");
+    cdata->codec = gst_v4l2_av1_get_codec ();
   } else if (gst_structure_has_name (s, "video/x-bayer")) {
     SET_META ("BAYER");
   } else if (gst_structure_has_name (s, "video/x-sonix")) {
@@ -1519,7 +1523,8 @@ gst_v4l2_video_dec_register (GstPlugin * plugin, const gchar * basename,
     }
 
     if (cdata->codec != NULL && cdata->codec != gst_v4l2_vp8_get_codec ()
-        && cdata->codec != gst_v4l2_vp9_get_codec ()) {
+        && cdata->codec != gst_v4l2_vp9_get_codec ()
+        && cdata->codec != gst_v4l2_av1_get_codec ()) {
       GValue value = G_VALUE_INIT;
 
       if (gst_v4l2_codec_probe_levels (cdata->codec, video_fd, &value)) {
diff --git a/sys/v4l2/meson.build b/sys/v4l2/meson.build
index 61a6f054fa..46de893edf 100644
--- a/sys/v4l2/meson.build
+++ b/sys/v4l2/meson.build
@@ -29,6 +29,7 @@ v4l2_sources = [
   'gstv4l2vp8enc.c',
   'gstv4l2vp9codec.c',
   'gstv4l2vp9enc.c',
+  'gstv4l2av1codec.c',
   'v4l2_calls.c',
   'v4l2-utils.c',
   'tuner.c',
